### PR TITLE
tests: Refactor spawn json command

### PR DESCRIPTION
**Pull request by serathius**
cc @chaochn47 @ahrtr

### CODE REVIEW

**Comment by chaochn47** [comment]
LGTM. 

The e2e test failures in the CI are limited to `lease`. 
```
         lease_test.go:312: 
            	Error Trace:	lease_test.go:312
            	            				execute.go:29
            	            				asm_386.s:1319
            	Error:      	Received unexpected error:
            	            	match not found. Set EXPECT_DEBUG for more info Err: read /dev/ptmx: input/output error, last lines:
            	            	{"level":"warn","ts":"2022-05-10T22:55:17.039Z","caller":"flags/flag.go:93","msg":"unrecognized environment variable","environment-variable":"ETCDCTL_API=3"}
            	            	{"cluster_id":238453183653593855,"member_id":14578408409545168728,"revision":1,"raft_term":2,"ID":6293921973442619141,"TTL":20,"Error":""}
            	Test:       	TestLeaseGrantRevoke/ClientTLS
```

The missing "header" line where spawn json cmd expects is due to 
https://github.com/etcd-io/etcd/blob/b7be91f98b7a7a6202ed9e9bc37b692d484585a8/client/v3/lease.go#L71-L75
the same as I called out in https://github.com/etcd-io/etcd/pull/14020#discussion_r868453327

File: tests/framework/e2e/etcdctl.go [hunk 1]
```diff
@@ -42,7 +42,8 @@ func (ctl *EtcdctlV3) DowngradeEnable(version string) error {
 }
 
 func (ctl *EtcdctlV3) Get(key string, o config.GetOptions) (*clientv3.GetResponse, error) {
-	args := ctl.cmdArgs()
+	resp := clientv3.GetResponse{}
+	args := []string{}
```
Comment (ahrtr) [inline tests/framework/e2e/etcdctl.go:46]: nit: `var args []string` is recommended in this case. I recall some go linters report warning on this.

File: tests/framework/e2e/etcdctl.go [hunk 2]
```diff
@@ -132,58 +128,25 @@ func (ctl *EtcdctlV3) Delete(key string, o config.DeleteOptions) (*clientv3.Dele
...
-	if err != nil {
-		return nil, err
-	}
-	var resp clientv3.DeleteResponse
-	line, err := cmd.Expect("header")
-	if err != nil {
-		return nil, err
-	}
-	err = json.Unmarshal([]byte(line), &resp)
+	err := ctl.spawnJsonCmd(&resp, args...)
```
Comment (ahrtr) [inline tests/framework/e2e/etcdctl.go:131]: nit: I would suggest to define resp (line 120) right above this line.

File: tests/framework/e2e/etcdctl.go [hunk 3]
```diff
@@ -132,58 +128,25 @@ func (ctl *EtcdctlV3) Delete(key string, o config.DeleteOptions) (*clientv3.Dele
...
-	}
-	var resp clientv3.MemberListResponse
-	line, err := cmd.Expect("header")
-	if err != nil {
-		return nil, err
-	}
-	err = json.Unmarshal([]byte(line), &resp)
+	resp := clientv3.MemberListResponse{}
+	err := ctl.spawnJsonCmd(&resp, "member", "list")
 	return &resp, err
```
Comment (ahrtr) [inline tests/framework/e2e/etcdctl.go:138]: nit: Does it make more sense to return `nil` response when `err != nil`?

File: tests/framework/e2e/etcdctl.go [hunk 4]
```diff
@@ -132,58 +128,25 @@ func (ctl *EtcdctlV3) Delete(key string, o config.DeleteOptions) (*clientv3.Dele
...
-	}
-	var resp clientv3.MemberAddResponse
-	line, err := cmd.Expect("header")
-	if err != nil {
-		return nil, err
-	}
-	err = json.Unmarshal([]byte(line), &resp)
+	resp := clientv3.MemberAddResponse{}
+	err := ctl.spawnJsonCmd(&resp, "member", "add", name, "--learner", "--peer-urls", strings.Join(peerAddrs, ","))
 	return &resp, err
```
Comment (ahrtr) [inline tests/framework/e2e/etcdctl.go:144]: nit: same as above

File: tests/framework/e2e/etcdctl.go [hunk 5]
```diff
@@ -132,58 +128,25 @@ func (ctl *EtcdctlV3) Delete(key string, o config.DeleteOptions) (*clientv3.Dele
...
-		return nil, err
-	}
-	var resp clientv3.MemberRemoveResponse
-	line, err := cmd.Expect("header")
-	if err != nil {
-		return nil, err
-	}
-	err = json.Unmarshal([]byte(line), &resp)
+	resp := clientv3.MemberRemoveResponse{}
+	err := ctl.spawnJsonCmd(&resp, "member", "remove", fmt.Sprintf("%x", id))
```
Comment (ahrtr) [inline tests/framework/e2e/etcdctl.go:149]: same as above

File: tests/framework/e2e/etcdctl.go [hunk 6]
```diff
@@ -216,36 +179,24 @@ func (ctl *EtcdctlV3) flags() map[string]string {
...
 		args = append(args, fmt.Sprintf("--command-timeout=%s", o.Timeout))
 	}
 	if o.Physical {
 		args = append(args, "--physical")
 	}
-
-	return nil, SpawnWithExpect(args, fmt.Sprintf("compacted revision %v", rev))
+	resp := clientv3.CompactResponse{}
+	err := ctl.spawnJsonCmd(&resp, args...)
+	return &resp, err
```
Comment (ahrtr) [inline tests/framework/e2e/etcdctl.go:191]: same as above

File: tests/framework/e2e/etcdctl.go [hunk 7]
```diff
@@ -216,36 +179,24 @@ func (ctl *EtcdctlV3) flags() map[string]string {
...
 		Status   *clientv3.StatusResponse
 	}
-	line, err := cmd.Expect("header")
-	if err != nil {
-		return nil, err
-	}
-	err = json.Unmarshal([]byte(line), &epStatus)
+	err := ctl.spawnJsonCmd(&epStatus, "endpoint", "status", "--endpoints", strings.Join(ctl.endpoints, ","))
 	if err != nil {
 		return nil, err
```
Comment (ahrtr) [inline tests/framework/e2e/etcdctl.go:201]: This is a nice example for my comments above.

File: tests/framework/e2e/etcdctl.go [hunk 8]
```diff
@@ -295,37 +234,14 @@ func (ctl *EtcdctlV3) Health() error {
...
-	}
-	var resp clientv3.LeaseGrantResponse
-	line, err := cmd.Expect("ID")
-	if err != nil {
-		return nil, err
-	}
-	err = json.Unmarshal([]byte(line), &resp)
+	resp := clientv3.LeaseGrantResponse{}
+	err := ctl.spawnJsonCmd(&resp, "lease", "grant", strconv.FormatInt(ttl, 10))
 	return &resp, err
```
Comment (ahrtr) [inline tests/framework/e2e/etcdctl.go:239]: same as above

File: tests/framework/e2e/etcdctl.go [hunk 9]
```diff
@@ -295,37 +234,14 @@ func (ctl *EtcdctlV3) Health() error {
...
-	}
-	var resp clientv3.LeaseTimeToLiveResponse
-	line, err := cmd.Expect("id")
-	if err != nil {
-		return nil, err
-	}
-	err = json.Unmarshal([]byte(line), &resp)
+	resp := clientv3.LeaseTimeToLiveResponse{}
+	err := ctl.spawnJsonCmd(&resp, "lease", "timetolive", strconv.FormatInt(int64(id), 16))
 	return &resp, err
```
Comment (ahrtr) [inline tests/framework/e2e/etcdctl.go:245]: same as above

File: tests/framework/e2e/etcdctl.go [hunk 10]
```diff
@@ -343,82 +259,32 @@ func (ctl *EtcdctlV3) Defragment(o config.DefragOption) error {
...
-	}
-	var resp clientv3.LeaseLeasesResponse
-	line, err := cmd.Expect("id")
-	if err != nil {
-		return nil, err
-	}
-	err = json.Unmarshal([]byte(line), &resp)
+	resp := clientv3.LeaseLeasesResponse{}
+	err := ctl.spawnJsonCmd(&resp, "lease", "list")
 	return &resp, err
```
Comment (ahrtr) [inline tests/framework/e2e/etcdctl.go:264]: same as above

File: tests/framework/e2e/etcdctl.go [hunk 11]
```diff
@@ -343,82 +259,32 @@ func (ctl *EtcdctlV3) Defragment(o config.DefragOption) error {
...
-	}
-	var resp clientv3.LeaseKeepAliveResponse
-	line, err := cmd.Expect("ID")
-	if err != nil {
-		return nil, err
-	}
-	err = json.Unmarshal([]byte(line), &resp)
+	resp := clientv3.LeaseKeepAliveResponse{}
+	err := ctl.spawnJsonCmd(&resp, "lease", "keep-alive", strconv.FormatInt(int64(id), 16), "--once")
 	return &resp, err
```
Comment (ahrtr) [inline tests/framework/e2e/etcdctl.go:270]: same as above

File: tests/framework/e2e/etcdctl.go [hunk 12]
```diff
@@ -343,82 +259,32 @@ func (ctl *EtcdctlV3) Defragment(o config.DefragOption) error {
...
-	}
-	var resp clientv3.LeaseRevokeResponse
-	line, err := cmd.Expect("header")
-	if err != nil {
-		return nil, err
-	}
-	err = json.Unmarshal([]byte(line), &resp)
+	resp := clientv3.LeaseRevokeResponse{}
+	err := ctl.spawnJsonCmd(&resp, "lease", "revoke", strconv.FormatInt(int64(id), 16))
 	return &resp, err
```
Comment (ahrtr) [inline tests/framework/e2e/etcdctl.go:276]: same as above

File: tests/framework/e2e/etcdctl.go [hunk 13]
```diff
@@ -343,82 +259,32 @@ func (ctl *EtcdctlV3) Defragment(o config.DefragOption) error {
...
-	}
-	var resp clientv3.AlarmResponse
-	line, err := ep.Expect("alarm")
-	if err != nil {
-		return nil, err
-	}
-	err = json.Unmarshal([]byte(line), &resp)
+	resp := clientv3.AlarmResponse{}
+	err := ctl.spawnJsonCmd(&resp, "alarm", "list")
 	return &resp, err
```
Comment (ahrtr) [inline tests/framework/e2e/etcdctl.go:282]: same as above

same comment for some places below.

**Comment by ahrtr** [comment]
I reran the failure pipelines 3 times, but always failing. Please make sure the failures are not caused by this PR or some recent PRs. 

If the lease issue (the fields are not in the "header") is the reason for the failure, then why some other PRs (i.e. [14022](https://github.com/etcd-io/etcd/pull/14022) ) are all green?

**Comment by chaochn47** [comment]
> If the lease issue (the fields are not under the "header") is the reason for the failure, then why some other PRs (i.e. https://github.com/etcd-io/etcd/pull/14022 ) are all green?

Because other PRs checks `id` existence in the header. https://github.com/etcd-io/etcd/blob/066e510731bb258ccdb7e64e49c14d46e58beb9e/tests/framework/e2e/etcdctl.go#L353

**Comment by ahrtr** [comment]
> > If the lease issue (the fields are not under the "header") is the reason for the failure, then why some other PRs (i.e. #14022 ) are all green?
> 
> Because other PRs checks `id` existence in the header.
> 
> https://github.com/etcd-io/etcd/blob/066e510731bb258ccdb7e64e49c14d46e58beb9e/tests/framework/e2e/etcdctl.go#L353

Thanks for the clarification. Then the failure is indeed caused by this PR. We can leave the `LeaseList` as it's (in the main branch) for now, and add a comment there. Once we refactor the lease in future, we can revisit the the `(ctl *EtcdctlV3) LeaseList() `again.

**Review by ptabor — APPROVED** [review]

**Review by ahrtr — APPROVED** [review]

**Pull request merged**